### PR TITLE
feat: add rpc protocol, client and server

### PR DESCRIPTION
**Pull request by rklaehn**
## Description

Add rpc protocol, client and server

## Breaking Changes

- net_protocol::Blobs::new_with_events takes an Endpoint

Compared to the old rpc in iroh, client::blobs::share is gone. That rpc call was always a bit questionable, it just takes a hash and adds the node data, and as such really has nothing to do with blobs as such.

## Notes & open questions

~~Merge this as is, or first try out the rpc refactorings?~~

This is now against the new quic-rpc. There are a number of types in net_protocol that could be made private with some effort, but I would prefer to do this in a subsequent PR. All fns on Blobs have been made private already though.

## Change checklist

- [x] Self-review.
- [x] Documentation updates following the [style guide](https://rust-lang.github.io/rfcs/1574-more-api-documentation-conventions.html#appendix-a-full-conventions-text), if relevant.
- [x] ~~Tests if relevant.~~
- [x] All breaking changes documented.

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -65,13 +73,18 @@ rcgen = "0.12.0"
...
 
 [features]
-default = ["fs-store"]
+default = ["fs-store", "rpc"]
 downloader = ["dep:parking_lot", "tokio-util/time", "dep:hashlink"]
 fs-store = ["dep:reflink-copy", "redb", "dep:redb_v1", "dep:tempfile"]
 metrics = ["iroh-metrics/metrics"]
 redb = ["dep:redb"]
+rpc = ["dep:quic-rpc", "dep:quic-rpc-derive", "dep:nested_enum_utils", "dep:strum", "dep:futures-util", "dep:ref-cast", "dep:portable-atomic", "dep:walkdir", "downloader"]
+ref-cast = ["dep:ref-cast"]
```
Comment (dignifiedquire) [inline Cargo.toml]: why is this its own feature?

File: Cargo.toml [hunk 2]
```diff
@@ -65,13 +73,18 @@ rcgen = "0.12.0"
...
+portable-atomic = ["dep:portable-atomic"]
```
Comment (dignifiedquire) [inline Cargo.toml]: why is this its own feature?

File: Cargo.toml [hunk 3]
```diff
@@ -65,13 +73,18 @@ rcgen = "0.12.0"
...
+walkdir = ["dep:walkdir"]
```
Comment (dignifiedquire) [inline Cargo.toml]: why is this its own feature?

File: src/net_protocol.rs [hunk 4]
```diff
@@ -149,6 +152,14 @@ impl<S: crate::store::Store> Blobs<S> {
         &self.store
     }
 
+    pub(crate) fn rt(&self) -> LocalPoolHandle {
```
Comment (dignifiedquire) [inline src/net_protocol.rs:155]: might be useful to make this public, not sure

File: src/net_protocol.rs [hunk 5]
```diff
@@ -149,6 +152,14 @@ impl<S: crate::store::Store> Blobs<S> {
         &self.store
     }
```
Comment (dignifiedquire) [inline src/net_protocol.rs:154]: I think down in this file there are a bunch of method which are public, only to make RPC work, would be good to check which of those you can now make private

File: src/rpc.rs [hunk 6]
```diff
@@ -0,0 +1,871 @@
...
+    {
+        use Request::*;
+        match msg {
+            Blobs(msg) => self.handle_blobs_request(msg, chan).await,
+            Tags(msg) => self.handle_tags_request(msg, chan).await,
+        }
+    }
+
+    /// Handle a tags request
+    pub async fn handle_tags_request<C>(
```
Comment (dignifiedquire) [inline src/rpc.rs:76]: does this need to be public?

File: src/rpc/client/blobs.rs [hunk 7]
```diff
@@ -0,0 +1,1795 @@
...
+use crate::rpc::proto::blobs::{
+    AddPathRequest, AddStreamRequest, AddStreamUpdate, BatchCreateRequest, BatchCreateResponse,
+    BlobStatusRequest, ConsistencyCheckRequest, CreateCollectionRequest, CreateCollectionResponse,
+    DeleteRequest, ExportRequest, ListIncompleteRequest, ListRequest, ReadAtRequest,
+    ReadAtResponse, ValidateRequest,
+};
+
+/// Iroh blobs client.
+#[derive(Debug, Clone)]
+pub struct Client<C = BoxedConnector<RpcService>> {
```
Comment (dignifiedquire) [inline src/rpc/client/blobs.rs:109]: why is this not repr transparent?

File: src/rpc.rs (hunk 6 above)
Comment (rklaehn) [reply src/rpc.rs:76]: nope

File: src/rpc/client/blobs.rs (hunk 7 above)
Comment (rklaehn) [reply src/rpc/client/blobs.rs:109]: I don't need it in this direction. But you never know... So going to add it.

File: src/net_protocol.rs (hunk 5 above)
Comment (rklaehn) [reply src/net_protocol.rs:154]: Making all of this private would be a bit of effort.

**Pull request merged**